### PR TITLE
Supervision order REST endpoints

### LINKED ISSUE #2878: Supervision Orders (JSPUI Only)
From release plan spreadsheet
Estimate from release plan: TBD
Expressing interest: none

May need DCAT feedback.  Unclear if this is used much.  

Described at: https://wiki.lyrasis.org/display/DSDOC6x/Functional+Overview#FunctionalOverview-SupervisionandCollaboration
See JSPUI: http://demo.dspace.org/jspui/dspace-admin/supervise

Description from 6.x JSPUI help menu:
> ## Supervision Orders
> Select Supervisors in the administration area and you are presented with three options:
> * Add a Supervision Order
> * View Current Supervision Orders
> * Clean Supervision Order Database
>The sections below detail how to deal with common supervision order tasks.
> _Note: The option to Clean Supervision Order Database should be used once in a while just to perform basic maintenance operations on the database. It is completely automatic and takes a very short time to execute._
>
> ### Viewing Current Supervision Orders
> Select View Current Supervision Orders and you are presented with a list of all supervision orders currently in effect. From this page you can modify the policies for the supervision order or remove the supervision order. (see the sections below).
>
>The details you have on this page allow you to see the name of the group doing the supervising, the author of the item being supervised and the title of that item.
>
>You may also add a supervision order from this page as per the section _Adding Supervision Orders_ below.
>
> ### Adding Supervision Orders
>Select _Add a Supervision Order_ from either the main supervision order page or the Current Supervision Orders page.
>
>To set a supervision order you need to define the group to do the supervising and the item in the workspace to be supervised. In addition, this page gives you the option to set a default set of policies for the supervising group to have over the item.
>
>Select the desired group from the pull down list of groups on this page; this box contains all the groups that currently exist on the system. Next select the WorkSpace item to be Supervised. This means clicking on the round select box on the right of the item in the listing you can see on this page, which shows all items currently in user workspaces.
>
>_Note: the ID that is in the left hand column of the item listing is the database id for the workspace item, and is there for the reference of system administrators._
>
>Once you have chosen the group and the workspace item to be connected you may also select a default policy from the Initial Policy Setting pull down box. The default policy options that this provides are:
>* None - The supervision order is created but the supervising group is not given any policies regarding the item to be supervised. This usually indicates that the administrator has decided to manually set policies for this supervision order.
>* Observer - The supervising group is given READ access to the item (but not to any bundles or bitstreams that already exist). Any new bundles or bitstreams inherit the supervision group's policy to permit READ operations.
>* Editor - The supervising group is given ADD, WRITE, and READ access to the item (but not any bundles or bitstreams that already exist). Any new bundles or bitstreams inherit the supervising groups's policy to permit ADD, WRITE and READ operations.
>Once happy with your selection click on Submit Supervision Order and your settings will be applied, and you will be returned to the supervision order main page.
>
>### Removing Supervision Orders
>Go to the View Current Supervision Orders page and select Remove from the supervision order that to wish to remove. You will be asked to confirm the removal of the order, upon confirmation of which it will be removed from the system.
>
>### Customising Supervision Order Policies
>Go to the View Current Supervision Orders page and select Policies for the supervision order for which you would like to customise the policies. You will be taken to the Policies for Item page. From here you may customise the policies for the particular item.
>
>In order to set policies for the supervisor group you should select, for example, Add New Policy for the item or any of its component parts, then select the supervisor group that you are wanting to configure from the given list of groups. You can then select the Action that you want to perform and hit Save Policy.

### PR DESCRIPTION
**Pull request by corrad82-4s**
## References
_Add references/links to any related issues or PRs. These may include:_
* Fixes #2878
* Related to DSpace/RestContract#208 

## Description
A rest endpoint according to what defined in REST Contract, accepting CRUD operations for supervision orders, with usage limited to Administrators has been exposed. Two new discovery configurations have been added: "otherworkspace" returning all Workspace item a user is a member of their supervision group, and "supervision" returning all workspace and workflow items, used by UI Workflow admin page. 

## Instructions for Reviewers

List of changes in this PR:
Several files have been added and / or updated, please refer to list of Changed files for a complete list.

Our suggestion is to test this PR together with its [angular part](https://github.com/DSpace/dspace-angular/pull/2019) 

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [ ] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by corrad82-4s** [comment]
We have updated this PR with an IT proving that when groups or items are deleted their related supervision orders are deleted as well

File: dspace/config/spring/api/discovery.xml [hunk 1]
```diff
@@ -894,6 +901,79 @@
         <property name="spellCheckEnabled" value="true"/>
     </bean>
 
+    <!--The other workspace configuration settings for discovery -->
+    <bean id="otherWorkspaceConfiguration"
```
Comment (tdonohue) [inline dspace/config/spring/api/discovery.xml:905]: What do we mean by an "otherWorkspaceConfiguration"?  Could we change the name here (the phrase "other workspace" is not at all clear)? How does this relate to or differ from the "supervisionConfiguration?

**Review by tdonohue — CHANGES_REQUESTED** [review]
@corrad82-4s : I gave this a code review today, and I'm not sure I understand why this seems like a complete "redesign" of Supervision Orders from the way they worked in 6.x?

It appears in this PR that Supervision Orders are no longer limited to WorkspaceItems (i.e. the submission process) as they were in 6.x.  They now appear on WorkflowItems as well, and are linked directly to the **Item** object itself (and I'm not seeing anyplace they are ever removed once the Item goes to "in_archive" status... and no tests that prove the Supervision order goes away once the Item is archived).

It also appears that there is no migration process for migrating older Supervision Orders (from 6.x) to 7.x.  All the 6.x code appears to be replaced, but it hasn't been removed from the codebase, and the data is never migrated.

I think this entire design needs more discussion as it's not what I anticipated.  It's no longer simply recreating 6.x JPSUI Supervision Orders... Instead, it's a new type of Supervision Order, and I don't understand how it's expected to work (as there's no clear documentation that I can find) or the use cases behind it.

**Comment by corrad82-4s** [comment]
Hi @tdonohue , 
thank you for your comments, we are reviewing what suggested and reported and will get back to you soon

File: dspace-api/src/main/java/org/dspace/supervision/SupervisionOrder.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Id
+    @Column(name = "id")
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "supervision_orders_seq")
+    @SequenceGenerator(name = "supervision_orders_seq", sequenceName = "supervision_orders_seq", allocationSize = 1)
+    private Integer id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "item_id")
+    private Item item;
```
[the remark this answers is not in the document]
Comment (corrad82-4s) [reply dspace-api/src/main/java/org/dspace/supervision/SupervisionOrder.java:44]: This mainly was to be more adherent to what set in contract, where we expect receiving an uuid when a supervision order is set. IMHO referencing the Item can help us in logic where we are retrieving supervision orders or executing discovery queries to retrieve supervised workspace items. Of course we must be sure that supervision order, once the item is archived, is removed.

[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/supervision/SupervisionOrder.java:44]: @corrad82-4s ; In that case, I recommend adding a test to verify that the supervision order is *deleted* when the item changes state.  It seems like it should be deleted once the WorkspaceItem is removed (i.e. once the WorkspaceItem transitions into a WorkflowItem).  As it's unclear to me why the supervision order should even remain once the Item progresses into the Workflow approval state.  But, if there's a reason to still have the supervision order on the `WorkflowItem`, then we should **minimally** add a test to verify the Supervision order is deleted once the Item is archived.

[the remark this answers is not in the document]
Comment (corrad82-4s) [reply dspace-api/src/main/java/org/dspace/supervision/SupervisionOrder.java:44]: @tdonohue : Yes sorry, the supervision order must be deleted once WorkspaceItem transitions into a WorkflowItem, not when the Item is archived

File: dspace-api/src/main/java/org/dspace/discovery/SolrServiceSupervisionOrderIndexingPlugin.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Autowired(required = true)
+    private SupervisionOrderService supervisionOrderService;
+
+    @Override
+    public void additionalIndex(Context context, IndexableObject indexableObject, SolrInputDocument document) {
+        try {
+
+            if (!(indexableObject instanceof IndexableWorkspaceItem) &&
+                !(indexableObject instanceof IndexableWorkflowItem)) {
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceSupervisionOrderIndexingPlugin.java:40]: I have reviewed our initial notes about that. The reason why we need to keep track of the supervision also when the item is in workflow is that workflow item can be rejected back to the workspaceitem. The idea is that if an item was supervised it should be still supervised when it goes back. In version 6 this was managed in an inconsistent way as the resource policies related to the supervision was left in place and so available when the item was sent back but the workspaceitem2eperson link was of course loss.
Please note that keep the information about the supervision when the item is in workflow add no functional changes as the supervisors cannot do anything to the workflowitem. This can be eventually proven by a dedicated test

File: dspace-api/src/main/java/org/dspace/supervision/SupervisionOrder.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/supervision/SupervisionOrder.java:44]: @corrad82-4s  @tdonohue  is right in saying that the supervisor order must be deleted once the item is archived but the supervisor order must be retained during the workflow for the reason explained here https://github.com/DSpace/DSpace/pull/8621/files#r1083021166 in my other reply

File: dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V7.5_2022.12.09__Supervision_Orders_table.sql [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+-------------------------------------------------------------------------------
+-- Table to store supervision orders
+-------------------------------------------------------------------------------
+
+CREATE TABLE supervision_orders
+(
+  id INTEGER PRIMARY KEY,
+  item_id UUID REFERENCES Item(uuid) ON DELETE CASCADE,
+  eperson_group_id UUID REFERENCES epersongroup(uuid) ON DELETE CASCADE
+);
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V7.5_2022.12.09__Supervision_Orders_table.sql:18]: The sql migration script is missing. In v6 we have the workspaceitem2eperson table that must be migrated to the new table. Moreover, checking the v6 code we have identified an issue, the resourcepolicies added for a supervisor order were never deleted, I suggest to create such policies with `TYPE_SUBMISSION` as by their nature are only relevant for the submission in the workspaceitem. In such way the policies will be automatically removed when the item is archived

File: dspace-api/src/main/java/org/dspace/app/exception/ResourceConflictException.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+package org.dspace.app.exception;
+
+/**
+ * This class provides an exception to be used when a conflict on a resource
+ * occurs.
+ *
+ * @author Mohamed Eskander (mohamed.eskander at 4science dot it)
+ *
+ */
+public class ResourceConflictException extends RuntimeException {
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/app/exception/ResourceConflictException.java:17]: not sure we really need that

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/exception/DSpaceApiExceptionControllerAdvice.java [hunk 6]
```diff
@@ -235,6 +248,12 @@ protected void handleGenericException(HttpServletRequest request, HttpServletRes
 
     }
 
+    @ExceptionHandler(ResourceConflictException.class)
+    protected ResponseEntity<? extends RestModel> resourceConflictException(ResourceConflictException ex) {
+        RestModel resource = converterService.toRest(ex.getResource(), utils.obtainProjection());
+        return new ResponseEntity<RestModel>(resource, HttpStatus.CONFLICT);
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/exception/DSpaceApiExceptionControllerAdvice.java:254]: I don't think we really need to introduce a new status here. This seems more related to 422 error code

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/model/WorkspaceItemRest.java [hunk 7]
```diff
@@ -14,10 +14,18 @@
  *
  * @author Andrea Bollini (andrea.bollini at 4science.it)
  */
+@LinksRest(links = {
+    @LinkRest(
+        name = WorkspaceItemRest.SUPERVISION_ORDERS,
+        method = "getSupervisionOrders"
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/model/WorkspaceItemRest.java:20]: see https://github.com/DSpace/DSpace/pull/8621/files#r1083021166

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SupervisionOrderRestRepository.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        }
+
+        Group group = groupService.find(context, UUID.fromString(groupId));
+        if (group == null) {
+            throw new UnprocessableEntityException("Group with uuid: " + groupId + " not found");
+        }
+
+        supervisionOrder = supervisionOrderService.findByItemAndGroup(context, item, group);
+        if (Objects.nonNull(supervisionOrder)) {
+            throw new ResourceConflictException(
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/SupervisionOrderRestRepository.java:126]: by previous agreement this special kind of conflict related to business rule was decided to be managed with a 422 see for instance here https://github.com/DSpace/RestContract/pull/180#discussion_r850829376

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java [hunk 9]
```diff
@@ -5987,4 +5991,491 @@ public void discoverFacetsTestWithDsoTypeTest() throws Exception {
...
+                                                          .withIssueDate("2010-11-03")
+                                                          .withTitle("Admin Workflow Item 1").build();
+
+        // create Four supervision orders for above items
+        SupervisionOrderBuilder.createSupervisionOrder(context, wsItem1Admin.getItem(), groupA).build();
+
+        SupervisionOrderBuilder.createSupervisionOrder(context, wsItem2Admin.getItem(), groupA).build();
+
+        SupervisionOrderBuilder.createSupervisionOrder(context, wfItem1Admin.getItem(), groupA).build();
+        SupervisionOrderBuilder.createSupervisionOrder(context, wfItem1Admin.getItem(), groupB).build();
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:6312]: I would like to support this use case at the least (initially) from the REST perspective.
An administrator can add a supervisor order to a workflowitem before reject it so that the supervisor can help in moving forward the submission

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/SupervisionOrderRestRepositoryIT.java [hunk 10]
```diff
@@ -0,0 +1,1088 @@
...
+        getClient(adminToken).perform(patch("/api/core/items/" + item.getID())
+                            .content(patchBody)
+                            .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$.uuid", Matchers.is(item.getID().toString())))
+                        .andExpect(jsonPath("$.withdrawn", Matchers.is(true)))
+                        .andExpect(jsonPath("$.inArchive", Matchers.is(false)));
+
+        getClient(adminToken).perform(get("/api/core/supervisionorders/" + supervisionOrder.getID()))
+                             .andExpect(status().isOk())
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/SupervisionOrderRestRepositoryIT.java:283]: I agree with @tdonohue this should lead to an exception of illegalstate. On the rest side this should be a 422. The item to supervise must be an inprogress submission (workspace/workflow)

File: dspace/config/spring/api/discovery.xml (hunk 1 above)
Comment (abollini) [reply dspace/config/spring/api/discovery.xml:905]: maybe supervisedWorkspaceConfiguration as there is already a supervionConfiguration used by the administrative dashboard?

**Review by abollini — CHANGES_REQUESTED** [review]
@corrad82-4s I have added my review to this PR as well. I take the chance to respond to some questions from @tdonohue about the scope/purpose of some decision, it looks to me that the code is almost ready but some cleanup and fixes are required

File: dspace-api/src/main/java/org/dspace/discovery/SolrServiceSupervisionOrderIndexingPlugin.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceSupervisionOrderIndexingPlugin.java:40]: @abollini : That makes sense as a use case for why a Supervision Order might remain on a `WorkflowItem`.  I can agree to that.

In that case, I'd recommend we create more tests which align with this use case / prove that it works as expected.  For instance, based on this use case, I'd assume the following should be true:
1. When an Item moves from WorkspaceItem to WorkflowItem, the SupervisionOrder is retained.  However, the users/groups related to the SupervisionOrder only have `READ` privileges on the Item when it is in Workflow (in the same way that a Submitter can only "read" their submission once it is in Workflow & can no longer modify it unless it is rejected & returned to "WorkspaceItem" status).  
    * In other words, while in WorkspaceItem status, a SupervisionOrder can allow for edit privileges. But, once in WorkflowItem status, it only allows for read privileges.
2. Obviously, when the Item becomes archived, the SupervisionOrder should be deleted.

**Comment by corrad82-4s** [comment]
@tdonohue , @abollini thank you for these reviews, they are all being addressed, updating the PR once ready

**Comment by corrad82-4s** [comment]
@tdonohue , @abollini we have addressed the feedback and updated the PR. It is ready for a review. In the next days. 
Please note that at least one of addressed feedback impacts on Angular PR, we will update it in the next days.

File: dspace-api/src/main/java/org/dspace/discovery/SolrServiceSupervisionOrderIndexingPlugin.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceSupervisionOrderIndexingPlugin.java:44]: Closing this request... I've verified these new tests have been added.

**Review by tdonohue — COMMENTED** [review]
@corrad82-4s : Just wanted to note that I've re-reviewed the code of this backend PR.  It all looks good to me now.  So, once the frontend bugs are resolved, I think the backend is ready to go.

**Review by abollini — APPROVED** [review]
thanks @corrad82-4s my previous feedback has been addressed, it looks good to me now

**Review by tdonohue — APPROVED** [review]
👍 Thanks @corrad82-4s !  My feedback has been addressed & I've verified this is working well with the updated frontend.
